File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> prev(n + 1, 0);
        vector<int> cur(n + 1, 0);

        // tabulation with space optemization
        for(int idx = n; idx > 0; idx--){
            for (int lci = 0; lci <= n; lci++){
                int take = 0;
                if (lci == 0 || nums[lci - 1] < nums[idx - 1]){
                    take = 1 + prev[idx];
                } 

                int notTake = prev[lci];
                cur[lci] = max(take, notTake);
            }

            prev = cur;
        }

        return prev[0];
    }

    int f(int idx, int lci, vector<int>& nums){
        if (idx > nums.size()) return 0;

        // take
        int take = 0;
        if (lci == 0 || nums[lci - 1] < nums[idx - 1]){
            take = 1 + f(idx + 1, idx, nums);
        }

        // not take
        int notTake = f(idx + 1, lci, nums);

        return max(take, notTake);
    }

};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int result = 0;
        int[][] dp = new int[n + 1][n + 2];


        for (int prev = n; prev >= 0; prev--) {
            for (int idx = n; idx >= 1; idx--) {
                int take = Integer.MIN_VALUE;

                if (prev == 0 || nums[prev - 1] < nums[idx - 1]) {
                    take = dp[idx][idx + 1] + 1;
                }

                int notTake = dp[prev][idx + 1];

                dp[prev][idx] = Math.max(take, notTake); 
            }
        }

        for (int i = 1; i <= n; i++) {
            result = Math.max(result, dp[0][i]);
        }

        return result;
    }


    


   
}